### ISSUE TITLE
Scope introspection builtins

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#492]
It would be nice if we have a pair of builtins for examining (lexical) scopes.

Scopes have two parts: a `names` part, storing variable names, and a `uses` part, storing other namespaces used in that scope:

https://github.com/elves/elvish/blob/2d837f121e850acdc3aeadec46929299dafd54ed/eval/scope.go#L3-L7

Hence it makes sense to have two builtin functions for examining these two parts. One `list-names` and one `list-uses`.

`list-names` takes an optional namespace name and lists all names in that namespace. For instance, `list-names builtin` lists all names in the builtin namespace, and `list-names local` lists all names in the local namespace. When no namespace or `""` is given, it should list all names in the local, up and builtin namespaces, akin to how variable name resolution works

https://github.com/elves/elvish/blob/2d837f121e850acdc3aeadec46929299dafd54ed/eval/eval.go#L368-L404

In short, "list-names *ns*" should list all such names *n* that $*ns*:*n* is a valid variable.

`list-uses` does not take arguments, and always lists names of namespaces used in the local, up, or builtin scope. This is to mirror the behavior of namespace resolution, which always happens as described:

https://github.com/elves/elvish/blob/2d837f121e850acdc3aeadec46929299dafd54ed/eval/eval.go#L406-L417

**Comment by krader1961** [comment]
Let's say you have namespaces `a:` and `a:b:` (from doing `use a` and `use a:b`). How would I use these commands to list just the namespaces nested within `a:`? Or is that a nonsensical question? It looks like command completions provide reasonable suggestions when I type `a:`<tab> including showing the nested `a:b:` namespace. Isn't that something we should be able to replicate using elvish script?

**Comment by xiaq** [comment]
@krader1961 There is no "nested namespace". The module `a:b:` happens to have `a:` as its prefix in name. (The situation is comparable to Perl's modules, if you are familiar with that). The fact that completion for `a:` includes `a:b:` is based on prefix matching, just like if you have variables `$a-b` and `$a`, typing `$a` will include `$a-b` as a candidate.

**Comment by krader1961** [comment]
Yeah, I see that now that I've read the docs more carefully after my previous question. The problem is that most people are going to look at namespaces a: and a:b: and conclude the latter is somehow nested inside the former. A careful reading of the rules for evaluating use statements makes the relationship clearer. But I suspect it will be a FAQ since nearly every new user will make the mistake I made. Which doesn't argue for changing the namespace model but does mean additional documentation, especially examples, is going to be needed.

**Comment by xiaq** [comment]
A different API has been implemented: namespaces can now be indexed like maps, and `has-key` and `keys` work on namespaces.

**Comment by zzamboni** [comment]
In my tests, `keys` works fine on namespaces, but `has-key` does not:
```
> kind-of $edit:
▶ ns
> keys $edit:
▶ move-dot-eol~
▶ rprompt-persistent
▶ match-prefix~
▶ complex-candidate~
▶ selected-file
...
> has-key $edit: after-readline
Exception: couldn't get key or index of type 'ns'
```
